src/evaluators/UnaryExpression.js

```diff
@@ -29,13 +31,124 @@ import {
 import type { BabelNodeUnaryExpression } from "babel-types";
 
 function isInstance(proto, Constructor): boolean {
   return proto instanceof Constructor || proto === Constructor.prototype;
 }
 
+function evaluateDeleteOperation(expr: Value | Reference, realm: Realm) {
+  // ECMA262 12.5.3.2
+
+  // 1. Let ref be the result of evaluating UnaryExpression.
+  let ref = expr;
+
+  // 2. ReturnIfAbrupt(ref).
+
+  // 3. If Type(ref) is not Reference, return true.
+  if (!(ref instanceof Reference)) return realm.intrinsics.true;
+
+  // 4. If IsUnresolvableReference(ref) is true, then
+  if (Environment.IsUnresolvableReference(realm, ref)) {
+    // a. Assert: IsStrictReference(ref) is false.
+    invariant(!Environment.IsStrictReference(realm, ref), "did not expect a strict reference");
+
+    // b. Return true.
+    return realm.intrinsics.true;
+  }
+
+  // 5. If IsPropertyReference(ref) is true, then
+  if (Environment.IsPropertyReference(realm, ref)) {
+    // a. If IsSuperReference(ref) is true, throw a ReferenceError exception.
+    if (Environment.IsSuperReference(realm, ref)) {
+      throw realm.createErrorThrowCompletion(realm.intrinsics.ReferenceError);
+    }
+
+    // b. Let baseObj be ! ToObject(GetBase(ref)).
+    let base = Environment.GetBase(realm, ref);
+    // Constructing the reference checks that base is coercible to an object hence
+    invariant(base instanceof ConcreteValue || base instanceof AbstractObjectValue);
+    let baseObj = base instanceof ConcreteValue ? To.ToObject(realm, base) : base;
+
+    // c. Let deleteStatus be ? baseObj.[[Delete]](GetReferencedName(ref)).
+    let deleteStatus = baseObj.$Delete(Environment.GetReferencedName(realm, ref));
+
+    // d. If deleteStatus is false and IsStrictReference(ref) is true, throw a TypeError exception.
+    if (!deleteStatus && Environment.IsStrictReference(realm, ref)) {
+      throw realm.createErrorThrowCompletion(realm.intrinsics.TypeError);
+    }
+
+    // e. Return deleteStatus.
+    return new BooleanValue(realm, deleteStatus);
+  }
+
+  // 6. Else ref is a Reference to an Environment Record binding,
+  // a. Let bindings be GetBase(ref).
+  let bindings = Environment.GetBase(realm, ref);
+  invariant(bindings instanceof EnvironmentRecord);
+
+  // b. Return ? bindings.DeleteBinding(GetReferencedName(ref)).
+  let referencedName = Environment.GetReferencedName(realm, ref);
+  invariant(typeof referencedName === "string");
+  return new BooleanValue(realm, bindings.DeleteBinding(referencedName));
+}
+
+function generateRuntimeCall(ref: Value | Reference, ast: BabelNodeUnaryExpression, strictCode: boolean, realm: Realm) {
+  let args = [];
+  let [thisArg, propName] = ref instanceof Reference ? [ref.base, ref.referencedName] : [];
+  if (thisArg instanceof Value) args = [thisArg];
+  if (propName !== undefined && typeof propName !== "string") args.push(propName);
+  if (thisArg instanceof Value) {
+    Leak.leakValue(realm, thisArg, ast.loc);
```

Comment:
When is leaking necessary in practice? Seems like the only case `delete` could trigger outside code is if this is a proxy. Otherwise I don't see a reason to leak.

How do you feel about skipping this if `thisArg` is a simple object?